### DISCUSSION TITLE
📁 Arrangement Of App Folders In Django

### DISCUSSION THREAD

**Discussion by taljacob2** [redhat-beyond/JobSeeker#47] [Ideas]
Hi all! 🥇 🏅 
All of our 5 "feature"-apps will make use of some "global" files. For example: `base_template.html`, `icons` folder `main.css` and more...

We should arrange our *Django Application* such that the `jobseeker`-app is a "global" app that is being injected into each "feature"-app. This way, we could make use of "global" **shared** files together in a modular way!

This is the *Django Application Rendering Flow* that will help us **share** data between our "feature"-apps:

[![](https://mermaid.ink/img/pako:eNqdVl1vmzAU_SuWpe6pYXmckFJpVTtp0h6mtttLmYoDJiED2zImU1T63-dr82USIAmVirk-516fa3PIO454TLGPb27Qw46wDUdfhcjSiKiUMx89URZTKVO2Qd8y_i9gif4fbYlU6OU-YL_JRhKm0GJxV_2UfJ8WmlWFa85VoSQRXrENzWQoUkHZHsmSIXFQW85QThjZUE8cIFhQuacSLT3z539ZLpfoU1jVawoY0ldRrnU1sa2DrwE-XnGA_1gwXHEqaQRR9OOpi4Zt4RCtVneoshqLigjRoZyCekZXC_UtRELync6KEp5pllOvuTTul8yKVyB8LvUIag1wuuZIsR1fF5T-pRJKtg8hZJ0q2iJt6fZxbAEO51np7kV9VmEiYV1xivtCc5ERRZ2iqgmOZBiXn1Aag3K4z4oGkNULoympMN-oNNhZgYDqaTOk62Xp1txzIuN6U9_W8HDOphpWu6eWN7OnhtLb0pp0zpYaqrujNft66UK_XJyRTBtEkmYUOtCE3oSNzTZikMP2Y5hlqi2DBE13jlLMNmmQqNero1zXt0w7rII-wX22NwCyDYHRVBNgvlFusLNqAdWTaEjXy8r4JmWgywxmhRmUVWaGU9IMoNFm0bPiDKynztIukteN9Ae0tb8uWn8K7IfmO4uyMqZF5Xh1wJwk4DNdwEHqL-min6ZxvxNw2wlN4ICHDxaNkeJVZ4MnSG0rTvPa6cGKW48YXbbnLQbyW1ebWLs3XIRrbdMCRsljKoZv75iYlbuVJ4xpXNJqNVjUSVeaFDaXYkwevLlnHqvGTy44Vp2xXHasHHsZrNi8jeceqdYqLjlPPc-48DC5zmHJxhcaDfgW51TmJI31b-t3iAVYbWlOA-zrYUwTUmYqwAH70NBSxDrPY5wqLrGfkKygt5iUij8fWIR9JUvagB5S-Mmd16iP_1mJyfI)](https://mermaid.live/edit#pako:eNqdVl1vmzAU_SuWpe6pYXmckFJpVTtp0h6mtttLmYoDJiED2zImU1T63-dr82USIAmVirk-516fa3PIO454TLGPb27Qw46wDUdfhcjSiKiUMx89URZTKVO2Qd8y_i9gif4fbYlU6OU-YL_JRhKm0GJxV_2UfJ8WmlWFa85VoSQRXrENzWQoUkHZHsmSIXFQW85QThjZUE8cIFhQuacSLT3z539ZLpfoU1jVawoY0ldRrnU1sa2DrwE-XnGA_1gwXHEqaQRR9OOpi4Zt4RCtVneoshqLigjRoZyCekZXC_UtRELync6KEp5pllOvuTTul8yKVyB8LvUIag1wuuZIsR1fF5T-pRJKtg8hZJ0q2iJt6fZxbAEO51np7kV9VmEiYV1xivtCc5ERRZ2iqgmOZBiXn1Aag3K4z4oGkNULoympMN-oNNhZgYDqaTOk62Xp1txzIuN6U9_W8HDOphpWu6eWN7OnhtLb0pp0zpYaqrujNft66UK_XJyRTBtEkmYUOtCE3oSNzTZikMP2Y5hlqi2DBE13jlLMNmmQqNero1zXt0w7rII-wX22NwCyDYHRVBNgvlFusLNqAdWTaEjXy8r4JmWgywxmhRmUVWaGU9IMoNFm0bPiDKynztIukteN9Ae0tb8uWn8K7IfmO4uyMqZF5Xh1wJwk4DNdwEHqL-min6ZxvxNw2wlN4ICHDxaNkeJVZ4MnSG0rTvPa6cGKW48YXbbnLQbyW1ebWLs3XIRrbdMCRsljKoZv75iYlbuVJ4xpXNJqNVjUSVeaFDaXYkwevLlnHqvGTy44Vp2xXHasHHsZrNi8jeceqdYqLjlPPc-48DC5zmHJxhcaDfgW51TmJI31b-t3iAVYbWlOA-zrYUwTUmYqwAH70NBSxDrPY5wqLrGfkKygt5iUij8fWIR9JUvagB5S-Mmd16iP_1mJyfI)

Practically said, the way to create this modular structure is by doing the following:

### Urls
- [x] `app/urls.py` should look like so:
   ```py
   from django.contrib import admin
   from django.urls import path, include
  
   urlpatterns = [
       path('admin/', admin.site.urls),
       path('', include('jobseeker.urls')),
   ]
   ```
- [x] `jobseeker/urls.py` should look like so:
   ```py
   from django.urls import path, include
   from . import views
  
   urlpatterns = [
       path('about/', views.about, name='about'),
       path('', include('feed.urls')),
       path('', include('job_board.urls')),
       path('', include('personal_profile.urls')),
       path('', include('chat.urls')),
       path('', include('login.urls')),
   ]
   ```
### Templates
- [x] Create a `jobseeker/templates` folder that contains:
   -  `base_template.html`.

### Static
- [x] Create a `jobseeker/static` folder that contains:
   - `main.css`
   - `main_logo.png`
   - `icons`
      - `facebook.svg`
      - `instagram.svg`
      - `twitter.svg`

--- 

- @paOmer Is this possible for you to do so in #20? Because we are all depending on that PR. I can help you do that.
- If you can't do that, another option is that I'll make the changes needed on another PR that is based on #20, and you will approve it along with our whole team.

**Comment by paOmer** [comment]
Hi @taljacob2,
The new arrangement that you suggest looks awesome.
I'll make it happen, but I’m thinking to make #20 a draft for now and do the new arrangement in a new PR.

**Comment by taljacob2** [comment]
> Hi @taljacob2,
> The new arrangement that you suggest looks awesome.
> I'll make it happen, but I’m thinking to make #20 a draft for now and do the new arrangement in a new PR.

Sounds great! :+1:

**Comment by taljacob2** [comment]
Thanks for @paOmer for excellent work! Each team member can now create his own "feature"-app based on #34.

**Comment by taljacob2** [comment]
@ShirleyFichman @paOmer @DeanBiton @amittcohen 

## GUIDE: How To Render The View In Your "Feature"-App

Say *YOUR_FEATURE_APP_NAME* is one of the following names:
- feed
- job_board
- personal_profile
- chat
- login

### Once you have created your "feature"-app:
1. Create a `templates/YOUR_FEATURE_APP_NAME` folder in your "feature"-app, and place there all your `.html` files.
    (For example: `YOUR_FEATURE_APP_NAME.html`)
1. Navigate to `app/settings.py`, and in line 60 add your templates folder to your "feature"-app you have created.
    For example: (added the line: `os.path.join(BASE_DIR, 'YOUR_FEATURE_APP_NAME', 'templates', 'YOUR_FEATURE_APP_NAME')`)
    ```py
        'DIRS': [os.path.join(BASE_DIR, 'templates'),
                 os.path.join(BASE_DIR, 'jobseeker', 'templates', 'jobseeker'),
                 os.path.join(BASE_DIR, 'YOUR_FEATURE_APP_NAME', 'templates', 'YOUR_FEATURE_APP_NAME')],
    ```
1. To render a `.html`, navigate to `YOUR_FEATURE_APP_NAME/views.py`, and make it look like so: (for example)
    ```py
    from django.shortcuts import render
    
    def YOUR_FEATURE_APP_NAME(request):
        return render(request, 'YOUR_FEATURE_APP_NAME.html')

    ```
1. Render your "feature"-app URLs by creating a `YOUR_FEATURE_APP_NAME/urls.py` file, with the following content:
    ```py
    from django.urls import path
    from . import views
    
    
    urlpatterns = [
        path('YOUR_FEATURE_APP_NAME/', views.YOUR_FEATURE_APP_NAME, name='YOUR_FEATURE_APP_NAME'),
    ]
    ```
1. Make sure `jobseeker/urls.py` is rendering your `YOUR_FEATURE_APP_NAME/urls.py`.
    You need to enable the `path` to `YOUR_FEATURE_APP_NAME/urls.py` in the `jobseeker/urls.py` file:
    ```py
    from django.urls import path, include
    from . import views
    
    
    urlpatterns = [
        path('about/', views.about, name='about'),
        path('', include('YOUR_FEATURE_APP_NAME.urls')),
    ]
    ```
1. You're done! The `.html` of your "feature"-app can be accessed at: http://localhost:8000/YOUR_FEATURE_APP_NAME/
